macro/src/expand.rs

```diff
@@ -859,6 +860,16 @@ fn expand_cxx_vector(namespace: &Namespace, elem: &Ident) -> TokenStream {
                 }
                 unsafe { __vector_size(v) }
             }
+            fn __vector_slice(v: &::cxx::CxxVector<Self>) -> &[Self] {
+                extern "C" {
+                    #[link_name = #link_size]
+                    fn __vector_size(_: &::cxx::CxxVector<#elem>) -> usize;
+                    #[link_name = #link_data]
+                    fn __vector_data(_: &::cxx::CxxVector<#elem>) -> *const #elem;
+                }
+                let ret = unsafe { std::slice::from_raw_parts(__vector_data(v), __vector_size(v)) };
```

Comment:
I believe this is undefined behavior in the case that size is 0. The data ptr of a &\[T\] has to be nonnull and properly aligned regardless of size, which C++ does not guarantee.

See https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html#safety.

Somewhere there will need to be special handling of the empty vector case.

Reply:
I was actually wondering about that. ;)
Let me add the special casing of nullptr/0 size

Reply:
Have not looked into how rust handles alignment. That could be a biggie if rust has different alignment reqs...

Having said that, we could enforce that in the __vector_data() implementation.

Reply:
Fixed handling of null data() ptr.

UB only happens if std::vector's capacity() is 0 if I got the from_raw_parts documentation right. There does not seem to be a need to specifically handle 0 size. Famous last words?

Ignore my going off  the tangent with non-alignment. I was thinking off T, whereas you were talking about the data ptr if null.